### PR TITLE
[#174917271] Upgrade Ruby and Rails

### PR DESCRIPTION
**Pull request by barrettclark**
# Rails 6 Upgrade

### Some initial cleanup

* Remove development.log file from project
* Update git ignore to current standard
* Remove Twilio references

### Upgrade

I created a new Rails 6 app so that I had access to all the right files:

```
rails new Rails6Project \
  --skip-action-cable \
  --skip-action-text \
  --skip-action-mailbox \
  --skip-action-mailer \
  --skip-active-record \
  --skip-active-storage \
  --skip-bundle \
  --skip-coffee \
  --webpack
```

There is a rake task to upgrade the project in place, but I like this better. Next we go through and clean up the old files:

* `rm -rf vendor test public log lib config bin`
* `mv app app_old`
* `cp -r ../Rails6Project/{app,bin,config,lib,log,public,test,vendor} .`
* Make sure the Rails6Project name is replaced throughout the project
* `cp app_old/controllers/main_controller app/controllers`
* `cp -r app_old/views/main app/views`
* `rm Gemfile.lock`
* `bundle install`

JavaScript/asset setup: https://dev.to/amree/rails-6-with-bootstrap-webpacker-for-js-asset-pipeline-for-css-2lmn
* `yarn add bootstrap jquery popper.js`
* `git mv app/assets/stylesheets/application.css app/assets/stylesheets/application.scss`
* remove the includes and add `@import 'bootstrap';`
* create the `app/assets/stylesheets/bootstrap.scss` file for all the specific bootstrap imports
* Add bootstrap import and setup in `app/javascript/packs/application.js`
* `rails webpacker:install`
* Then update `config/webpack/environment.js` to make jQuery available

And then you should be able to run the upgraded Rails 6 app

### So What _Is_ This App?

This is a very simple application. There are no models -- not unless the applicant adds any. There is a controller and a view. Everything else in the project is a stock Rails app with Bootstrap added.

### Rubocop Cleanup

We don't want to have `tcs-rubocop` in the Gemfile because it is not publicly accessible. So instead I included the `rubocop` and `rubocop-performance` gems and pinned them to the versions we current have in `tcs-rubocop` and copied the default yml file into the project for the rubocop config. With that in place I ran `rubocop -a`.

(#174917271)

### CODE REVIEW

**Review by jldalton — COMMENTED** [review]
Can't say I follow all the changes, but I understand a lot of this is due to the Rails upgrade and Rubocop. Really just have one thing to consider in the instructions.

File: config/initializers/content_security_policy.rb [hunk 1]
```diff
@@ -1,3 +1,4 @@
+# frozen_string_literal: true
```
[the remark this answers is not in the document]
Comment (barrettclark) [reply config/initializers/content_security_policy.rb:1]: This is a Ruby convention that started a few versions ago in preparation for Ruby 3

File: README.md [hunk 2]
```diff
@@ -1,2 +1,11 @@
 # Programming-Exercise-Rails
+
 Project for Ruby/Rails developer candidates
+
+## Setup Instructions
+
+1. Be sure you have Ruby 2.6.5 installed.
```
[the remark this answers is not in the document]
Comment (barrettclark) [reply README.md:7]: Good point. "at least" is better. I don't think 2.7.x is too different, and this app is simple enough that it shouldn't really matter. When Ruby 3 drops at Christmas we may need to revisit this.

[the remark this answers is not in the document]
Comment (barrettclark) [reply README.md]: I updated the second commit and force pushed

File: config/initializers/content_security_policy.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (jldalton) [reply config/initializers/content_security_policy.rb:1]: Not blocking this PR, but what's the latest on this? I read on one web page that Matz may have reconsidered making all strings frozen by default in v3?

[the remark this answers is not in the document]
Comment (jldalton) [reply config/initializers/content_security_policy.rb:1]: Here is the page I read: https://bugs.ruby-lang.org/issues/11473#note-53

**Review by jldalton — APPROVED** [review]

File: config/initializers/content_security_policy.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (barrettclark) [reply config/initializers/content_security_policy.rb:1]: I don't know how I missed that this conversation had occurred a year or so ago. Very interesting - thank you for sharing it. This is something that we should discuss in the Ruby SIG, and possibly modify our [tcs-]rubocop styleguide config.

[the remark this answers is not in the document]
Comment (petesharum) [reply config/initializers/content_security_policy.rb:1]: This is news to me too - it is still included in [the default rubocop config](https://github.com/rubocop-hq/rubocop/blob/master/config/default.yml#L3085). Considering it isn't likely to become default behavior any time soon, I'm in favor of reviewing our style guide accordingly.

File: Gemfile [hunk 3]
```diff
@@ -1,17 +1,56 @@
...
+end
+
+# Windows does not include zoneinfo files, so bundle the tzinfo-data gem
+gem "tzinfo-data", platforms: [:mingw, :mswin, :x64_mingw, :jruby]
+
+# TCS additions
+gem "haml-rails"
+
+# Mirror the tcs-rubocop version
+gem "rubocop", "0.85.0"
```
Comment (petesharum) [inline Gemfile:55]: Should we include gems like rubocop in the programming exercise? I'm not concerned with testing a candidate's knowledge of individual gems, and sometimes the point of a programming exercise is to present the candidate with code that is intentionally bad.

File: config/initializers/content_security_policy.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (barrettclark) [reply config/initializers/content_security_policy.rb:1]: There is some performance benefit from reduced garbage collection, but having this in the styleguide gets a comment/question more often than not when all the files get updated. I've added it to the Ruby SIG things to discuss.

File: Gemfile (hunk 3 above)
Comment (barrettclark) [reply Gemfile:55]: My thinking with this was twofold:

1. We should have the project represent us as closely as possible, and our styleguide is part of that.
  1.1. Does that mean that the gem should be included?
  1.2. Does that mean that the code should be updated with out without the gem being included?
2. Having the code represent our style is one thing, and hopefully the candidate picks up on that and mimics it. This is part of our rubric. Having the gem and styleguide in the project without making any mention of it also acts as a breadcrumb.

As I write these points out and consider what you've said I don't really know where I stand. How does the Java exercise handle this? I'm fine with pulling that third commit out and pushing the upgraded project forward. We can always come back and add the cleanup.

**Pull request merged**